From 834b91cd9bf364eaa9fedf7f3bd142b0e7ee1740 Mon Sep 17 00:00:00 2001
From: "Luo, Ying" <ying2.luo@intel.com>
Date: Mon, 9 Mar 2020 12:46:33 +0800
Subject: [PATCH] Modify OmafDashAccess CMakeLists for android support

Signed-off-by: Luo, Ying <ying2.luo@intel.com>
---
 OmafDashAccess/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/OmafDashAccess/CMakeLists.txt b/OmafDashAccess/CMakeLists.txt
index 3fdb2040..dc5f960c 100644
--- a/OmafDashAccess/CMakeLists.txt
+++ b/OmafDashAccess/CMakeLists.txt
@@ -7,7 +7,7 @@ AUX_SOURCE_DIRECTORY(../utils UTIL_SRC)
 AUX_SOURCE_DIRECTORY(OmafDashDownload/. DOWN_SRC)
 AUX_SOURCE_DIRECTORY(OmafDashParser/. PARSE_SRC)

-ADD_DEFINITIONS("-g -c -fPIC -lstdc++fs -std=c++11 -D_GLIBCXX_USE_CXX11_ABI=0 -z noexecstack -z relro -z now -fstack-protector-strong -fPIE -fPIC -pie -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Wl,-S -Wall -Werror")
+ADD_DEFINITIONS("-g -c -fPIC -std=c++11 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIE -fPIC -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Wall -Werror")

 INCLUDE_DIRECTORIES(/usr/local/include ../utils ../isolib)
 LINK_DIRECTORIES(/usr/local/lib ../isolib/dash_parser)
--
2.17.1
